libvast/src/system/evaluator.cpp

```diff
@@ -0,0 +1,173 @@
+                                            const ids& result) {
+  VAST_DEBUG(self, "got new hits", result, " for predicate", pred);
+  if (auto ptr = sub_hits_of(pred); ptr == nullptr) {
+    VAST_ERROR(self, "got a result for an unknown predicate:", pred);
+  } else {
+    auto& [missing, accumulated_hits] = *ptr;
+    accumulated_hits |= result;
+    if (--missing == 0) {
+      VAST_DEBUG(self, "collected all INDEXER results for predicate", pred);
+      update();
+    }
+  }
+  decrement_pending();
+}
+
+void evaluator_state::handle_missing_indexer_result(const predicate& pred,
+                                                    const caf::error& err) {
+  VAST_IGNORE_UNUSED(err);
+  VAST_DEBUG(self, "got no result for predicate", pred, "INDEXER returned",
+             self->system().render(err));
+  if (auto ptr = sub_hits_of(pred); ptr == nullptr) {
+    VAST_ERROR(self, "got an error for an unknown predicate:", pred);
+  } else if (--ptr->first == 0) {
+    VAST_DEBUG(self, "collected all INDEXER results for predicate", pred);
+    update();
+  }
+  decrement_pending();
+}
+
+void evaluator_state::update() {
+  VAST_DEBUG(self, "got sub_hits:", sub_hits,
+             "expr_hits:", caf::visit(ids_evaluator{sub_hits}, expr));
+  auto delta = caf::visit(ids_evaluator{sub_hits}, expr) - hits;
+  if (any<1>(delta)) {
+    hits |= delta;
+    self->send(client, std::move(delta));
+  }
+}
+
+void evaluator_state::decrement_pending() {
+  // We're done evaluating if all INDEXER actors have reported their hits.
+  if (--pending_responses == 0) {
+    VAST_DEBUG(self, "completed expression evaluation");
+    self->send(client, done_atom::value);
+  }
+}
+
+evaluator_state::sub_hits_map::mapped_type*
+evaluator_state::sub_hits_of(const predicate& pred) {
+  auto i = sub_hits.find(pred);
+  return i != sub_hits.end() ? &i->second : nullptr;
+}
+
+caf::behavior evaluator(caf::stateful_actor<evaluator_state>* self,
+                        std::vector<caf::actor> indexers) {
+  // Skip incoming queries when already processing one.
+  return {[=](const expression& expr, caf::actor client) {
+    self->state.init(client, expr);
+    // TODO: we might want to locate the smallest subset of indexers (checking
+    //       whether the predicate could match the type of the indexer) instead
+    //       of always querying all indexers.
+    auto predicates = caf::visit(predicatizer{}, expr);
+    if (predicates.empty()) {
+      VAST_DEBUG(self, "could not generate any predicates from expression");
+      self->send(client, done_atom::value);
+      return;
+    }
+    self->state.pending_responses = predicates.size() * indexers.size();
+    for (auto& pred : predicates) {
+      self->state.sub_hits.emplace(pred, std::pair{indexers.size(), ids{}});
+      for (auto& x : indexers)
+        self->request(x, caf::infinite, pred)
+          .then(
+            [=](const ids& hits) {
+              // Forward to state.
```

Comment:
What do you mean with this comment? That the implementation is delegated to `state`? If so, I think it's obvious and we can remove the comment. (Same with next handler.)